data_simulation.py

```diff
@@ -0,0 +1,223 @@
+    # GT states
+    states = generate_gt_state(A, number_of_nucleotids)
+    strands = evolution(X, states, trees, Q)
+
+    return strands, states
+
+
+def scale_branches_length(tree, scale=0.1):
+    '''Given a tree in the dictionnary of list format it returns another
+    tree with randomised branches length
+        Args:
+            - tree (dict): the tree referencing the relationships between nodes
+            and the branches length.
+            - max_amp (float, optionnal) : maximal amplitude of branch length
+        Output:
+            - tree with same shape but randomised branches length
+    '''
+    tree_cp = deepcopy(tree)
+
+    def rescale_node(node):
+        childs = tree_cp[node]
+        if childs:
+            for child in childs:
+                new_child = child["node"]
+                rescale_node(new_child)
+                child["branch"] *= scale
+
+    rescale_node(max(tree_cp.keys()))
+    return tree_cp
+
+
+def rate_sub_HKY(pi, kappa):
+    """ define the rate substitution matrices according to the HKY model for
+    all states
+    Args :
+           - pi (np vector, size alphabetSize) transition probabilities
+           - kappa (np vector, size nb states) translation transversion rate
+    returns : Q (np array, nb states x alphabetSize x alphabetSize)
+    the rate substituon matrices for a states
+    """
+    nbState = len(kappa)
+    alphabetSize = len(pi)
+    Q = np.zeros((nbState, alphabetSize, alphabetSize))
+    for j in range(nbState):
+        for i in range(alphabetSize):
+            Q[j, i, :] = pi[i]
+            Q[j, i, (i + 2) % alphabetSize] *= kappa[j]
+            # put in each diagonal a term such that the rows sums to 0
+            Q[j, i, i] -= np.sum(Q[j, i, :])
+    return Q
+
+
+def generate_initial_vector(b, nbNucleotids):
+    '''Return a random vector of nucleotids as integers
+        Args:
+            - b (array of float, sums to 1) initial discrete distribution
+            - nbNucleotids (int), size of the output
+        Output:
+            - np.vector with values between 0 and size(b)-1
+            follows distribution b
+    '''
+    cumsum = np.cumsum(b)
+    random_values = np.random.rand(nbNucleotids)
+    X = np.empty(nbNucleotids, dtype=np.uint8)
+    for i in range(b.shape[0]):
+        X[random_values < cumsum[i]] = i
+        random_values[random_values < cumsum[i]] = 1
+    return X
+
+
+def generate_gt_state(A, nbNucleotids):
+    '''Use the state transition matrix A to generate of state path
+        Args:
+            - nbState (np matrix) state transition matrix
+            - nbNucleotids (int) length of the DNA in Nucleotids
+        Output:
+            - np. vector of int from 0 to nbState-1
+    '''
+    states = np.empty(nbNucleotids, dtype=np.uint8)
+
+    nbState = A.shape[0]
+    states[0] = np.random.randint(0, nbState, 1)[0]
+    for i in range(nbNucleotids-1):
+
+        discrete_law = A[states[i]]
```

Comment:
en python ces deux choses sont équivalentes. 
à chaque itération je cherche à déterminer l'état suivant *i+1*